### DISCUSSION TITLE
Packaging modules

### DISCUSSION THREAD

**Discussion by Cromwell1963** [pyinstaller/pyinstaller#7445] [Help]
I have written a python program with number of dynamically imported modules.  My hope is to release the main program and release additional modules later. This way I hope not to have to redistribute the main code every time it is added to. Is this possible using PyInstaller? If so, could someone give some pointers as to how? perhaps some relevant documentation. thanks

**Comment by rokm** [comment]
In theory it is possible; the extra plugin modules (.py source files, or .pyc files that you manually byte-compiled) can be imported from any location in `sys.path`. But it is not recommended unless you really know what you are doing.

For example, the subsequently released plugin modules should not introduce additional dependencies, because the main application's embedded python environment will contain only the python modules (stdlib and 3rd party) that were available at original build time and were required by the code that was analyzed at that time. If your original plugin modules are truly dynamically/indirectly loaded, they might not be analyzed at all, and you'll need to force analysis via `hiddenimports`.

And if you wanted to somehow include/load and external 3rd party package (e.g., additional dependency of some later plugin module), you need to be careful that none of that package was already collected in main application - because the (partially) collected copy in application's executable-embedded archive will shadow the external copy. The only exemption to that is if you add it to application's run-time directory (pointed to by `sys._MEIPASS` at runtime). Which is possible for `onedir` builds but more difficult for `onefile` ones (where you would probably need to somehow copy extra stuff into temporary directory at the very start of your program).

**Comment by Cromwell1963** [comment]
Thank for you for your quick and helpful answer. I was aware of the need to use `hiddeninports`, and have taken note of your comments about dependencies and 3rd party packages. At this stage I am pleased to know that it is possible, but am certainly open to alternatives. thanks!
